orchagent/dtelorch.cpp

```diff
@@ -0,0 +1,1384 @@
+
+    string& report_session_id = m_dtelEventTable[event].reportSessionId;
+
+    decreaseReportSessionRefCount(report_session_id);
+
+    m_dtelEventTable.erase(event);
+}
+
+sai_status_t DTelOrch::updateSinkPortList()
+{
+    sai_attribute_t attr;
+    attr.id = SAI_DTEL_ATTR_SINK_PORT_LIST;
+    sai_status_t status = SAI_STATUS_SUCCESS;
+
+    vector<sai_object_id_t> port_list;
+
+    for (auto it = sinkPortList.begin(); it != sinkPortList.end(); it++)
+    {
+        if (it->second != 0)
+        {
+            port_list.push_back(it->second);
+        }
+    }
+
+    attr.value.objlist.count = (uint32_t)port_list.size();
+    if (port_list.size() == 0)
+    {
+        attr.value.objlist.list = {};   
+    } else {
+        attr.value.objlist.list = port_list.data();
+    }
+    
+    status = sai_dtel_api->set_dtel_attribute(dtelId, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("DTEL ERROR: Failed to set INT sink port list");
+        return status;
+    }
+
+    return status;
+}
+
+bool DTelOrch::addSinkPortToCache(const string &port_alias, const sai_object_id_t port_id)
+{
+    if (sinkPortList.find(port_alias) == sinkPortList.end())
+    {
+        return false;
+    }
+
+    sinkPortList[port_alias] = port_id;
+    return true;
+}
+
+bool DTelOrch::removeSinkPortFromCache(const string &port_alias)
+{
+    if (sinkPortList.find(port_alias) == sinkPortList.end())
+    {
+        return false;
+    }
+
+    sinkPortList[port_alias] = 0;
+    return true;
+}
+
+void DTelOrch::update(SubjectType type, void *cntx)
+{
+    sai_status_t status;
+
+    if (type != SUBJECT_TYPE_PORT_CHANGE)
+    {
+        return;
+    }
+
+    PortUpdate *update = static_cast<PortUpdate *>(cntx);
+
+    /* Check if sink ports need to be updated */
+    if (update->add)
+    {
+        if (addSinkPortToCache(update->port.m_alias, update->port.m_port_id))
+        {
+            status = updateSinkPortList();
+            if (status != SAI_STATUS_SUCCESS)
+            {
+                SWSS_LOG_ERROR("DTEL ERROR: Failed to update sink port list on port add");
+                return;
+            }
+        }
+    } else {
+        if (removeSinkPortFromCache(update->port.m_alias))
+        {
+            status = updateSinkPortList();
+            if (status != SAI_STATUS_SUCCESS)
+            {
+                SWSS_LOG_ERROR("DTEL ERROR: Failed to update sink port list on port remove");
+                return;
+            }
+        }
+    }
+
+    /* Check if queue reports need to be updated */
+    auto port_entry_iter = m_dTelPortTable.find(update->port.m_alias);
+
+    if (port_entry_iter == m_dTelPortTable.end())
+    {
+        return;
+    }
+
+    dTelPortQueueTable_t qTable = port_entry_iter->second.queueTable;
+
+    for (auto it = qTable.begin(); it != qTable.end(); it++ )
+    {
+        DTelQueueReportEntry qreport = it->second;
+        if (update->add)
+        {
+            if (qreport.queueReportOid != 0)
```

Comment:
Hi,
Besides the port creation/deletion, does the queue can be created/deleted dynamically(I am not sure about this, I appreciate if you can help provide the answer)? If so do we also need to update the queue report table according to the queue creation/deletion?

Reply:
As far as I see in swss code, only port creation/deletion results in new OID allocation for the respective queues

Reply:
I doubt the assumption hold here. whether or not queue id changes during port creation and deletion depends on SAI implementation. I doubt SAI has this assumption. To be on the safe side, once port is created, the queue id should be re-queried and added.

similiar to above port update logic, whether a queue should be reported or not should come from user input. My assumption here is that if a port is newly create, the port should come up clean, meaning has only default configuration, in this case sink port, queue report should be explicitly added by user. We should not implicitly add them to sink port or queue report.

Reply:
maybe that is true for your implementation, but certainly it is not the assumption for SAI. The object id is opaque to the user, each SAI implementation can decide how to assign object id as long as it is unique.

Reply:
We are not exposing the queue OIDs to the users. Users configure using relative queue numbers for the port: HMSET DTEL_QUEUE_REPORT|Ethernet8|0 ...
Here 0 is the first queue for Ethernet8. DtelOrch agent will query portsorch agent to get the actual OID for this queue and then configure the queue report.